instance/tasks.py

```diff
@@ -339,3 +344,109 @@ def cleanup_old_betatest_users():
         users_to_delete = queryset.filter(is_active=False, profile__modified__lte=delete_cutoff)
         logger.info('Deleting %s users.', users_to_delete.count())
         users_to_delete.delete()
+
+
+class KillZombiesRunner:
+    """
+    Helper class to run `kill_zombies_periodically`
+    """
+
+    def __init__(self):
+        self.region: str = settings.OPENSTACK_REGION
+        self.recipient: list = [email for _, email in settings.ADMINS]
+        self.threshold: int = settings.KILL_ZOMBIES_WARNING_THRESHOLD
+
+    def send_email(self, subject: str, body: str) -> int:
+        """
+        Utility method for sending emails
+        """
+        if self.recipient:
+            email = EmailMessage(
+                subject, body, settings.DEFAULT_FROM_EMAIL, self.recipient
+            )
+            email.send()
+        else:
+            logging.info("Email recipient is undefined. Skipping email.")
+
+    def trigger_warning(self, num_zombies):
+        """
+        Warns when more than `threshold` VMs will be terminated
+        """
+        subject = "Zombie instances are over the current threshold ({})".format(
+            self.threshold
+        )
+        body = (
+            "The number of zombie OpenStack VMs ({}) in the {} region "
+            "is over the KILL_ZOMBIES_WARNING_THRESHOLD ({}).\n"
+        ).format(num_zombies, self.region, self.threshold)
+        logging.info(
+            "%s\nSending an email notification to: %s",
+            body, self.recipient
+        )
+        self.send_email(subject, body)
+
+    def get_zombie_servers_count(self, stdout: str) -> int:
+        """
+        If there are servers to terminate, a `kill_zombies` dry run will warn:
+        "Would have terminated {} zombies" to stdout.
+        """
+        num_instances = 0
+        if "No servers found in region" in stdout:
+            return num_instances
+        try:
+            num_instances = int(stdout.split("Would have terminated")[1].split()[0])
+        except (IndexError, ValueError):
+            logger.info("Received unexpected input from dry run. Defaulting to zero")
+        return num_instances
+
+    def run(self):
+        """
+        Main method for running the task
+        """
+        try:
+            dry_run_output, output = "", ""
+            dry_run_output = call_command(
+                "kill_zombies",
+                region=self.region,
+                dry_run=True,
+            )
+            num_of_zombies = self.get_zombie_servers_count(dry_run_output)
+            if num_of_zombies > self.threshold:
+                self.trigger_warning(num_of_zombies)
+            elif num_of_zombies != 0:
+                output = call_command("kill_zombies", region=self.region)
+            logging.info(
```

Comment:
@spokerman12, shouldn't this log statement be indented to be within the `elif` block so that the log message is printed only when `num_of_zombies != 0`?

Reply:
It could, but at this point, even if `num_of_zombies != 0`, the task did run successfully and this is the log entry for saying so. If I put it inside the logging call I'd have to write another logging call. Logging `task run successfully and there were no VMs to terminate` is a valid and desirable output to me.
Do you suggest otherwise?

Reply:
@spokerman12, okay to leave this as-is to avoid another commit and the following CI checks.

Reply:
@spokerman12, I spoke too soon. Your PR branch is already outdated and has to be updated. So you can go ahead and add the additional log statement.